### DISCUSSION TITLE
Get WI type wise with associated details

### DISCUSSION THREAD

**Discussion by JaydeepUniverse** [rfennell/AzurePipelines#1082] [Q&A] [answered]
I have opened a new discussion as recommended

From my original question posted here https://github.com/rfennell/AzurePipelines/issues/890#issuecomment-919969028

I need to have other details associated with WI as well like Assigned to, Tags, Parents in table format

For example, I tried as below but AssignedTo is not working, so just want help here

```
#### 🐞 Bugs
| WI ID | Tags | Assigned To |
|--|--|--|
{{#forEach this.workItems}}
{{#with fields}}
{{#if (eq (get 'System.WorkItemType' this) 'Bug')}}
| #{{../id}} | {{'System.Tags'}} | {{#with (lookup 'System.AssignedTo')}} {{displayName}} {{/with}} |
{{/if}}
{{/with}}
{{/forEach}}
```

And how can I use parent in same table?

```
#### 🐞 Bugs
| WI ID | Tags | Assigned To | Parent |
|--|--|--|--|
{{#forEach this.workItems}}
{{#with fields}}
{{#if (eq (get 'System.WorkItemType' this) 'Bug')}}
| #{{../id}} | {{'System.Tags'}} | {{#with (lookup 'System.AssignedTo')}} {{displayName}} {{/with}} | {{#return_parents_only this.workItems this.relatedWorkItems}} #{{this.id}} {{/return_parents_only}}  |
{{/if}}
{{/with}}
{{/forEach}}
```

As suggested by @rfennell for both above question, as below, 

```
#### 🐞 Bugs
| WI ID | Tags | Assigned To | Parent |
|--|--|--|--|
{{#forEach this.workItems}}
{{#if (eq (get 'System.WorkItemType' this) 'Bug')}}
| #{{this.id}} |  {{lookup this.fields 'System.Tags'}} |  {{#with (lookup this.fields 'System.AssignedTo')}} {{displayName}} {{/with}} |  {{#forEach this.relations}}
{{#if (contains this.attributes.name 'Parent')}}
{{#with (lookup_a_work_item ../../relatedWorkItems  this.url)}}
      - {{this.id}} - {{lookup this.fields 'System.Title'}}
      {{#forEach this.relations}}
      {{#if (contains this.attributes.name 'Parent')}}
      {{#with (lookup_a_work_item ../../../../relatedWorkItems  this.url)}}
         - {{this.id}} - {{lookup this.fields 'System.Title'}}
      {{/with}}
      {{/if}}
      {{/forEach}}
{{/with}}
{{/if}}
{{/forEach}}
  |
{{/if}}
{{/forEach}}
```

I am still getting blank output, which I think is due to WITH wrapper is not used for "fields" and from the output of this.workItems it is directly trying to find this.id but actually its like `fields.id`

`{{#if (eq (get 'System.WorkItemType' this) 'Bug')}}`

Please advise further.

Thanks

**Comment by rfennell** [comment, marked as answer]
I have tested this template and it works

```
| WI ID | Tags | Assigned To | Parent |
|--|--|--|--|
{{#forEach this.workItems}}
{{#if (eq (lookup this.fields 'System.WorkItemType') 'Bug')}}
| #{{this.id}} |  {{lookup this.fields 'System.Tags'}} |  {{#with (lookup this.fields 'System.AssignedTo')}} {{displayName}} {{/with}} |  {{#forEach this.relations}}{{#if (contains this.attributes.name 'Parent')}}{{#with (lookup_a_work_item ../../relatedWorkItems  this.url)}}#{{this.id}} - {{lookup this.fields 'System.Title'}}{{/with}}{{/if}}{{/forEach}} |
{{/if}}
{{/forEach}}
````
I get

![image](https://user-images.githubusercontent.com/324718/133475433-ae6e4e8f-6e99-4092-bc3c-d4b7b9bf54b5.png)

Points to note

- the task parameter `getParentsAndChildren` must be set to `true`
- you can't have new lines or white space within the table template to make it look neat, if you do it it get rendered out and ruins the table

**Comment by JaydeepUniverse** [comment]
Thanks @rfennell this new template is working now.

Continuing with parent WI query, now I am getting parent ID of WI but if I'd like to next level of parent of parent WI of WI then how could I get those details?

I see there is one task parameter
`getAllParents | Get All Parents for associated work items, recursing back to workitems with no parents e.g. up to Epics, defaults to false
`
I have selected this param in the task but I am getting one level of parent item only.

Do I need to make any changes in the handlebar template? please advise.

**Reply by rfennell** [reply]
Yes you are correct you need to enable the getallparents parameter.  If you do this all parents (up to a work item with no parent) to the array of related items

You also need to edit the template to add the loops to look for parents of parents. This needs to be repeated for all the levels you are expect, but the pattern is the same as the sample I provided

- get list of related items
- if a parent get the parent wi
- display what you want
- and report...

**Reply by JaydeepUniverse** [reply]
Thanks @rfennell after enabling getallparents parameter, I have looped again to get parent within parent loop and I am getting up level parent WIs.

**Comment by JaydeepUniverse** [comment]
Further, As you have mentioned already that if output contains new line or space then it ruins the table, I tested already before and its same, however any solution so far or through js extension is tried and doable?

**Reply by rfennell** [reply]
The table format issue is more a pain than a problem. The way I work is to get the logic working, with the Handlebar blocks all laid out for ease of reading. Once it is outputting the fields I want I then just removed spaces and line feeds to sort out the table. A sort of manual minify.

When doing this I do find using my [local tester](https://github.com/rfennell/AzurePipelines/tree/main/Extensions/XplatGenerateReleaseNotes/V3/testconsole) really speeds up this process. It allows you to rerun the note generation over and over again while you sort a template out.

As to using a Javascript handlbars extension, it is certainly an option. I have provided some [samples](https://github.com/rfennell/AzurePipelines/tree/main/SampleTemplates/XplatGenerateReleaseNotes%20(Node%20based)/Version%203) of extensions I have used. However, it is down to you to decide how much custom object reprocessing you want as opposed to Handlebar layout. But it should be possible

**Reply by JaydeepUniverse** [reply]
Ok thanks @rfennell I would try if feasible however for now we're proceeding with List format.

Many thanks for the help and extension which makes life easier.
